Fix Emoji UnicodeAndNames throwing exception because of duplicit keys

Rutherther opened the pull request:
Simple convert of key to value and vice versa when obtaining UnicodeAndNames from NamesAndUnicodes is not possible as there are multiple values to send the same emoji (:-), :), :slight_smile: etc.)

I've came up with a solution where the dictionary contains readonly collection of the possible names. Using LINQ the collection is grouped, then converted to dictionary.

Rutherther committed:
Fix Emoji UnicodeAndNames throwing exception because of multiple same keys

INikonI commented:
Given that the names are not used in any way, it is useless, it is better to change name `UnicodeAndNames` to `Unicodes` and type to `IReadOnlySet<string>` and make setter like this
```cs
private static IReadOnlySet<string> _unicodes;
private static IReadOnlySet<string> Unicodes
{
    get
    {
        _unicodes ??= NamesAndUnicodes.Select(kvp => kvp.Value).ToImmutableHashSet();
        return _unicodes;
    }
}
```
Also we can add `.ToImmutableDictionary()` to initalization of `NamesAndUnicodes`
```cs
private static IReadOnlyDictionary<string, string> NamesAndUnicodes { get; } = new Dictionary<string, string>
{
    ...
}.ToImmutableDictionary();
```

Rutherther committed:
Use ToImmutableDictionary instead of new ReadOnlyDictionary for UnicodesAndNames

Rutherther committed:
Use HashSet for Emoji Unicodes parse matching

Rutherther commented:
It's true that they aren't used anywhere, I have chosen this approach, because it was a `Dictionary` in the first place and I didn't want to drop out any features that may be used in the future. I don't know if there were any plans with this.

I added `Unicodes` property as you suggested and `ToImmutableDictionary` to avoid creating it using new, it's shorter now. I also changed `TryParse` to use `Unicodes` to prevent creating the whole dictionary (even though it's cached so I don't think there would be big performance issue).

I've kept `UnicodesAndNames`, still not sure whether it was intended for use later. You may deleted or I will if you guys want that.

INikonI commented:
Im added `UnicodesAndNames` just because was needed mirror of `NamesAndUnicodes`

INikonI commented:
I don't think that it will be particularly needed in the future, at least now only unicode is used in the api
